File: src/main/java/thecudster/sre/events/UpdateChecker.java
/*
 * SkyblockReinvented - Hypixel Skyblock Improvement Modification for Minecraft
 *  Copyright (C) 2021 theCudster
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published
 *  by the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
package thecudster.sre.events;

import com.google.gson.JsonObject;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.event.ClickEvent;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.EnumChatFormatting;
import net.minecraftforge.event.entity.EntityJoinWorldEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.versioning.DefaultArtifactVersion;
import thecudster.sre.SkyblockReinvented;
import thecudster.sre.util.api.APIHandler;

public class UpdateChecker {
    public boolean updateChecked = false;
    /*
    * Modified from Danker's Skyblock Mod under GNU General Public License.
    * https://github.com/bowser0000/SkyblockMod/blob/master/LICENSE
    * @author bowser0000
     */
    @SubscribeEvent
    public void check(EntityJoinWorldEvent event) {
        if (!updateChecked) {
            updateChecked = true;
            new Thread(() -> {
                EntityPlayer player = Minecraft.getMinecraft().thePlayer;

                System.out.println("Checking for updates...");
                JsonObject latestRelease = APIHandler.getResponse("https://api.github.com/repos/theCudster/SkyblockReinvented/releases");
                String latestTag = latestRelease.get("tag_name").getAsString();
                DefaultArtifactVersion currentVersion = new DefaultArtifactVersion(SkyblockReinvented.VERSION);
                DefaultArtifactVersion latestVersion = new DefaultArtifactVersion(latestTag.substring(1));

                if (currentVersion.compareTo(latestVersion) < 0) {
                    String releaseURL = latestRelease.get("html_url").getAsString();

                        ChatComponentText update = new ChatComponentText(EnumChatFormatting.GREEN + "" + EnumChatFormatting.BOLD + "  [UPDATE]  ");
                        update.setChatStyle(update.getChatStyle().setChatClickEvent(new ClickEvent(ClickEvent.Action.OPEN_URL, releaseURL)));

                        try {
                            Thread.sleep(2000);
                        } catch (InterruptedException ex) {
                            ex.printStackTrace();
                        }
                        player.addChatMessage(new ChatComponentText(SkyblockReinvented.VERSION + " is outdated. Please update to " + latestTag + ".\n").appendSibling(update));
                    }
                }).start();
            }
        }
}

File: src/main/java/thecudster/sre/core/UpdateChecker.kt
package thecudster.sre.core;

import com.google.gson.JsonArray;
import net.minecraft.client.gui.GuiMainMenu;
import net.minecraftforge.client.event.GuiOpenEvent;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import thecudster.sre.SkyblockReinvented;
import thecudster.sre.core.gui.UpdateCheckerGUI;
import thecudster.sre.util.api.APIUtil;

public class UpdateChecker {
    public String newVersionExists() {
        JsonArray apiObj = APIUtil.getJSONResponse("https://api.github.com/repos/theCudster/SkyblockReinvented/releases").getAsJsonArray();
        String version = apiObj.get(0).getAsJsonObject().get("tag_name").getAsString();
        if (!version.contains("v")) return null;
        version = version.substring(version.indexOf("v") + 1);
        if (version.contains("pre")) {
            double numVersion = Double.parseDouble(version.substring(0, version.indexOf("pre")));
            double preVersion = Double.parseDouble(version.substring(version.indexOf("pre") + 3));
            if (numVersion == myNumVersion() && preVersion == myPreVersion()) {
                return null;
            }
        } else {
            double numVersion = Double.parseDouble(version);
            if (numVersion == myNumVersion()) {
                return null;
            }
        }
        return version;
    }
    public double myNumVersion() {
        if (SkyblockReinvented.VERSION.contains("pre")) {
            return Double.parseDouble(SkyblockReinvented.VERSION.substring(0, SkyblockReinvented.VERSION.indexOf("-")));
        }
        return Double.parseDouble(SkyblockReinvented.VERSION);
    }
    public double myPreVersion() {
        if (!SkyblockReinvented.VERSION.contains("pre")) {
            return 0;
        }
        return Double.parseDouble(SkyblockReinvented.VERSION.substring(SkyblockReinvented.VERSION.indexOf("pre") + 3));
    }
    @SubscribeEvent(priority = EventPriority.HIGHEST)
    public void mainMenuOpen(GuiOpenEvent event) {
        if (event.gui instanceof GuiMainMenu) {
            if (newVersionExists() != null) {
                SkyblockReinvented.currentGui = new UpdateCheckerGUI();
            }
        }
    }
}